pkg/routine/recurringtask.go

```diff
@@ -47,34 +47,34 @@ func NewRecurringTask(t Task, i time.Duration, ops ...RecurringTaskOption) *Recu
 }
 
 // Start starts the timer
-func (t *RecurringTask) Start(ctx context.Context) error {
+func (t *RecurringTask) Start(_ context.Context) error {
 	t.ticker = t.clock.Ticker(t.interval)
 	ready := make(chan struct{})
 	go func() {
 		close(ready)
 		for {
 			select {
-			// TODO (soy) we can not cancel on ctx.Done, seems there is something cause context timeout of recurring task unexpected
-			case <-t.ch:
+			case <-t.done:
 				return
 			case <-t.ticker.C:
 				t.t()
 			}
 		}
 	}()
-
+	// ensure the goroutine has been running
 	<-ready
 	return t.TurnOn()
 }
 
 // Stop stops the timer
 func (t *RecurringTask) Stop(_ context.Context) error {
+	// prevent stop is called before start.
 	if err := t.TurnOff(); err != nil {
```

Comment:
not related to this PR, but the public function `TurnOff` called in `Stop` is weird.